In [15]:
# ============================================================
# 03_train_rf_classifier.ipynb
# Random Forest Model Training
# Breast Tumor Classification Project
# ============================================================

In [16]:
# ============================================================
# 1. Import Libraries
# ============================================================

import sys
from pathlib import Path

import pandas as pd
import joblib

# Add project root to Python path
sys.path.append(str(Path().resolve().parent))

In [17]:
# ============================================================
# 2. Import Custom Modules
# ============================================================

from src.split import stratified_split

from src.train import (
    create_random_forest_model,
    train_model,
    generate_predictions
)

In [18]:
# ============================================================
# 3. Load Processed Dataset
# ============================================================

df = pd.read_csv(
    "../data/processed/processed_data.csv"
)

df.head()

,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave_points_mean,symmetry_mean,fractal_dimension_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave_points_worst,symmetry_worst,fractal_dimension_worst,Diagnosis
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,1
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,1
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,1
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,1
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,1


In [19]:
# ============================================================
# 4. Split Features and Target
# ============================================================

X = df.drop(columns=['Diagnosis'])

y = df['Diagnosis']

print("Feature Matrix Shape:", X.shape)
print("Target Vector Shape:", y.shape)


Feature Matrix Shape: (569, 30)
Target Vector Shape: (569,)


In [20]:
# ============================================================
# 5. Perform Stratified Split
# ============================================================

(
    X_train,
    X_val,
    X_test,
    y_train,
    y_val,
    y_test
) = stratified_split(X, y)

In [21]:
# ============================================================
# 6. Print Split Shapes
# ============================================================

print("Training Set Shape:", X_train.shape)
print("Validation Set Shape:", X_val.shape)
print("Test Set Shape:", X_test.shape)

Training Set Shape: (398, 30)
Validation Set Shape: (85, 30)
Test Set Shape: (86, 30)


In [22]:
# ============================================================
# 7. Check Class Distribution
# ============================================================

print("\nTraining Distribution:")
print(y_train.value_counts(normalize=True))

print("\nValidation Distribution:")
print(y_val.value_counts(normalize=True))

print("\nTest Distribution:")
print(y_test.value_counts(normalize=True))


Training Distribution:
Diagnosis
0    0.628141
1    0.371859
Name: proportion, dtype: float64

Validation Distribution:
Diagnosis
0    0.623529
1    0.376471
Name: proportion, dtype: float64

Test Distribution:
Diagnosis
0    0.627907
1    0.372093
Name: proportion, dtype: float64


In [23]:
# ============================================================
# 8. Create Random Forest Model
# ============================================================

model = create_random_forest_model()

print(model)

RandomForestClassifier(n_estimators=200, random_state=42)


In [24]:
# ============================================================
# 9. Train Model
# ============================================================

model = train_model(
    model,
    X_train,
    y_train
)

print("Model training completed.")


Model training completed.


In [25]:
# ============================================================
# 10. Generate Validation Predictions
# ============================================================

val_predictions, val_probabilities = generate_predictions(
    model,
    X_val
)

print("Validation predictions generated.")

Validation predictions generated.


In [26]:
# ============================================================
# 11. Generate Test Predictions
# ============================================================

test_predictions, test_probabilities = generate_predictions(
    model,
    X_test
)

print("Test predictions generated.")

Test predictions generated.


In [27]:
# ============================================================
# 12. Save Trained Model
# ============================================================

model_path = "../artifacts/models/rf_model.joblib"

joblib.dump(
    model,
    model_path
)

print(f"Model saved at: {model_path}")

Model saved at: ../artifacts/models/rf_model.joblib


# Training Summary

## Model Training Workflow

1. The processed dataset was loaded successfully.

2. The dataset was split into:
   - Training Set (70%)
   - Validation Set (15%)
   - Test Set (15%)

3. Stratified splitting was used to preserve class distribution across all datasets.

4. A Random Forest Classifier was initialized using a fixed random state for reproducibility.

5. The model was trained using the training dataset.

6. Predictions and probability scores were generated for:
   - Validation dataset
   - Test dataset

7. The trained model was saved for future evaluation and inference tasks.

## Why These Steps Matter

- Train/Validation/Test separation prevents data leakage.
- Stratification ensures balanced class representation.
- Validation data supports unbiased tuning and threshold selection.
- Probability scores are required for ROC-AUC and threshold analysis.
- Saving the trained model improves reproducibility and deployment readiness.